src/storage/core.rs

```diff
@@ -933,6 +980,44 @@ where
     }
 }
 
+fn filter_deleted(all_entries: Vec<Entry>) -> Vec<Entry> {
+    let mut last_removal_by_key = HashMap::new();
+    let mut last_created_by_key = HashMap::new();
```

Comment:
This function is only used from `read_all`, which returns records for one particular key. So you don't need a map for different keys. Additionally, you can enforce this by adding that key as the argument.